File: examples/example_pipeline_provenance.ts
/**
 * pipeline_provenance_node.cpp — see examples/example_pipeline_provenance.py
 */
const sst = require('../index.js');

function baseRecord(extra) {
    return Object.assign(
        {
            stage: 0, // KnotPlot
            certification: 1, // Candidate
            toolName: 'tool',
            toolVersion: '1',
            inputSha256: 'in',
            outputSha256: 'out',
            parameterSha256: 'par',
            parentRecordSha256: '',
            coordinateConvention: 'xyz',
            scaleConvention: 'R=1',
            timestampUtc: '2026-01-01T00:00:00Z',
        },
        extra || {}
    );
}

function main() {
    if (!sst.isAvailable || typeof sst.evaluateProvenanceChain !== 'function') {
        console.log('[SKIP] pipeline_provenance: addon not available');
        return;
    }
    const a = baseRecord({ inputSha256: 'a0', outputSha256: 'a1', parentRecordSha256: '' });
    const fp = sst.provenanceRecordFingerprint(a);
    console.log('provenanceRecordFingerprint:', fp);

    const b = baseRecord({
        stage: 3, // SSTcore
        inputSha256: 'a1',
        outputSha256: 'a2',
        parentRecordSha256: fp,
    });
    const status = sst.evaluateProvenanceChain([a, b]);
    console.log('evaluateProvenanceChain:', status);
}

main();

File: examples/example_pipeline_provenance.py
"""Smoke example for pipeline_provenance bindings (Canon v0.8.23+)."""
from __future__ import annotations

import hashlib
import os
import sys

sys.path.insert(0, os.path.abspath("."))
sys.path.insert(0, os.path.join(os.path.abspath("."), "src"))

try:
    from SSTcore import (
        PipelineProvenanceAPI,
        ProvenanceRecord,
        PipelineStage,
        PipelineCertificationStatus,
    )
except ImportError:
    from sstcore import (
        PipelineProvenanceAPI,
        ProvenanceRecord,
        PipelineStage,
        PipelineCertificationStatus,
    )


def _h(label: str) -> str:
    return hashlib.sha256(label.encode("utf-8")).hexdigest()


def base(**kwargs):
    r = ProvenanceRecord()
    r.tool_name = "tool"
    r.tool_version = "1"
    r.input_sha256 = _h("in")
    r.output_sha256 = _h("out")
    r.parameter_sha256 = _h("par")
    r.coordinate_convention = "xyz"
    r.scale_convention = "R=1"
    r.timestamp_utc = "2026-01-01T00:00:00Z"
    for k, v in kwargs.items():
        setattr(r, k, v)
    return r


a = base(stage=PipelineStage.KnotPlot, input_sha256=_h("a0"), output_sha256=_h("a1"), parent_record_sha256="")
fp = PipelineProvenanceAPI.record_fingerprint(a)
print("record_fingerprint:", fp)

b = base(
    stage=PipelineStage.SSTcore,
    input_sha256=_h("a1"),
    output_sha256=_h("a2"),
    parent_record_sha256=fp,
)
status = PipelineProvenanceAPI.evaluate_chain([a, b])
print("evaluate_chain:", status)
print("Certified enum:", PipelineCertificationStatus.Certified)
print("stage_name:", PipelineProvenanceAPI.stage_name(PipelineStage.KnotPlot))
